### PR TITLE
SRCH-3908 redact PII from analytics log lines

### PR DESCRIPTION
**Pull request by MothOnMars**
## Summary
This PR is intended to redact PII such as email addresses and SSNs from the search and click log lines that are read by Logstash and indexed into Elasticsearch. Follow-up stories will redact that information from additional log lines.
 
### Checklist
Please ensure you have addressed all concerns below before marking a PR "ready for review" or before requesting a re-review. If you cannot complete an item below, replace the checkbox with the ⚠️ `:warning:` emoji and explain why the step was not completed.
 
#### Functionality Checks
 
- [x] You have merged the latest changes from the target branch (usually `main`) into your branch.
  
- [x] Your primary commit message is of the format **SRCH-#### \<description\>** matching the associated Jira ticket.

- [x] PR title is either of the format **SRCH-#### \<description\>** matching the associated Jira ticket (i.e. "SRCH-123 implement feature X"), or **Release - SRCH-####, SRCH-####, SRCH-####** matching the Jira ticket numbers in the release.
 
- [x] Automated checks pass. If Code Climate checks do not pass, explain reason for failures:
 
#### Process Checks

- [x] You have specified at least one "Reviewer".

### CODE REVIEW

File: lib/redactor.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  PATTERNS = {
+    ssn: /\b\d{3}[-\sxX._]?\d{2}[-\sxX._]?\d{4}\b/,
+    email: /[a-zA-Z0-9][-a-zA-Z0-9._]+@[-a-zA-Z0-9]+\.+[a-zA-Z]{2,8}/,
+    cc: /\b(?:\d[ -.]*?){13,17}\b/,
+    phone: /\(?([2-9]\d{2})\)?[- .]?([2-9]\d{2}[- .]?\d{4})/
+  }.freeze
+
+  def self.redact(str)
+    str = str.dup
+    PATTERNS.each { |key, value| str.gsub!(value, "[redacted_#{key}]") }
```
Comment (MothOnMars) [inline lib/redactor.rb:16]: I tried hard to find a more elegant way to do this, but it gets the job done. I confirmed that the performance impact is negligible when compared to raw logging:
```ruby
> Benchmark.realtime { 10000.times { Rails.logger.info("[Search Impression] #{Redactor.redact(hash.to_json)}") } }
0.8479499999666587
> Benchmark.realtime { 10000.times { Rails.logger.info("[Search Impression] #{hash.to_json}") } }
0.7438500000862405
```

**Review by lorawoodford — DISMISSED** [review]
@MothOnMars Thanks for the thorough discussion here/documentation in the `Redactor` module.  I agree, done is better than perfect at the moment.

LGTM

**Comment by MothOnMars** [comment]
@lorawoodford , this is ready for re-review. The changes address issues found during testing in staging, including:
- URI-encoded characters in PII (such as the "@" in email addresses)
- a too-hungry approach to log line redaction, which resulted in some information being redacted from non-sensitive log components such as user agents.

I wanted to do some additional DRY-ing up of the code, but held off from refactoring at this point. The `SearchImpression#log` and `Click#log` methods are similar but diverge in ways that make refactoring a bit tricky (`SearchImpression` is a module, while `Click` is a class). I'd like to make those more consistent, but this wasn't the story to address that.

File: spec/lib/redactor_spec.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+describe Redactor do
+  describe '.' do
```
Comment (lorawoodford) [inline spec/lib/redactor_spec.rb:4]: Perhaps

```suggestion
  describe '.redact' do
```

**Review by lorawoodford — APPROVED** [review]
One extremely-minor suggestion in line, otherwise approving/LGTM

File: spec/lib/redactor_spec.rb (hunk 2 above)
Comment (MothOnMars) [reply spec/lib/redactor_spec.rb:4]: Ack - that was a typo. Fixing...

**Review by lorawoodford — APPROVED** [review]
LGTM

**Pull request merged**